Create garbage collection docs (#28870)
kubernetes/website: content/en/docs/reference/glossary/finalizer.md

---
title: Finalizer
id: finalizer
date: 2021-07-07
full_link: /docs/concepts/overview/working-with-objects/finalizers/
short_description: >
  A finalizer is a resource key that specifies what actions to take before
  deleting specific resources.

aka: 
tags:
- fundamental
- operation
---

> A resource key that specifies what actions to take before deleting specific resources.

Author: Is it a field in the API or defined on a resource? (or both?)

> A resource key that specifies what actions to take before deleting specific resources.

Reviewer: It's a common field across multiple resources. See https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/object-meta/#System

<!--more-->

> 

Reviewer: ```suggestion
When you tell Kubernetes to delete an object, the Kubernetes API gives an
opportunity for the control plane (or some other API client) to take actions before
the deletion is considered as complete. For example, you can define a finalizer to
clean up related resources or infrastructure.

```

---
title: Finalizer
id: finalizer
date: 2021-07-07
full_link: /docs/concepts/overview/working-with-objects/finalizers/
short_description: >

>   A field in the Kubernetes API that specifies what actions to take

Reviewer 2: This short description does not seem clear -- especially the start of the sentence,
"A field in the Kubernetes API". There are a lot of fields ....
Compare to:
In Kubernetes, finalizers are keys that allow you to specify pre-delete
operations for resources.

Author: Would "A Kubernetes API resource that specifies.." work better? It was originally "A resource key that specifies" but we changed it in [this review comment](https://github.com/kubernetes/website/pull/28870#discussion_r666544260)

Reviewer: Finalizers aren't fields nor are they resources with URIs.
To me they are _identifiers_, namespaced similar to the keys for labels and annotations. AIUI: Kubernetes itself pays no heed to naming conventions for finalizers. The API server simply waits for the finalizer list to become empty as each defined cleanup mechanism completes.

Reviewer: For more help on this, maybe ask in [#wg-api-expression](https://kubernetes.slack.com/messages/wg-api-expression) on Kubernetes Slack.

Author: [Asked!](https://kubernetes.slack.com/archives/C0123CNN8F3/p1626183692078400)

Author: I changed the definition a bit to make it clearer (hopefully). @sftim PTAL. If it's OK, I have no other changes for this PR.

  before deleting specific resources. 

aka: 
tags:
- fundamental
- operation
---

> A finalizer is a field in the Kubernetes API that specifies what actions to take before
> deleting specific resources. 

Reviewer: How about:
```suggestion
Finalizers define a way to tell Kubernetes to hold off before fully deleting an object.
The main use of finalizers is to clean up any resources that deleted object owns.
```

<!--more-->

When you tell Kubernetes to delete an object, the Kubernetes API allows the
control plane (or another API client) to take specific actions before it
considers the deletion complete. These actions are defined by the finalizers

> defined for the resource.

Reviewer: ```suggestion
defined for the resource. The control plane (or other actor) marks that it has finished cleaning
up by removing its own finalizer from that object.
```
